### DISCUSSION TITLE
RPM 4.20.0 building problem

### DISCUSSION THREAD

**Discussion by iseki0** [rpm-software-management/rpm#3430] [Q&A] [answered]
When I trying to build following the `INSTALL` file, a linking problem blocks me.
I'm tried modify the `CMakeLists.txt` `tools/CMakeLists.txt`, but it doesn't works.

My commands:
```shell
mkdir _build
cd _build
cmake  .. -DLUA_LIBRARIES=/root/lua-5.4.7/src/liblua.a -DLUA_INCLUDE_DIR=../../lua-5.4.7/src -DWITH_SEQUOIA=OFF -DWITH_LEGACY_OPENPGP=ON
make
```
The lua was compiled with a flag `-fPIC`(I put it to the Makefile).

My problem:
```
[ 64%] Building C object tools/CMakeFiles/elfdeps.dir/elfdeps.c.o        
[ 65%] Linking C executable elfdeps             
/usr/bin/ld: ../rpmio/librpmio.so.10.2.0: undefined reference to `sinh'  
/usr/bin/ld: ../rpmio/librpmio.so.10.2.0: undefined reference to `log2'  
/usr/bin/ld: ../rpmio/librpmio.so.10.2.0: undefined reference to `atan2' 
/usr/bin/ld: ../rpmio/librpmio.so.10.2.0: undefined reference to `tanh'  
/usr/bin/ld: ../rpmio/librpmio.so.10.2.0: undefined reference to `cosh'  
/usr/bin/ld: ../rpmio/librpmio.so.10.2.0: undefined reference to `fmod'  
/usr/bin/ld: ../rpmio/librpmio.so.10.2.0: undefined reference to `acos'  
/usr/bin/ld: ../rpmio/librpmio.so.10.2.0: undefined reference to `sin'   
/usr/bin/ld: ../rpmio/librpmio.so.10.2.0: undefined reference to `asin'  
/usr/bin/ld: ../rpmio/librpmio.so.10.2.0: undefined reference to `exp'   
/usr/bin/ld: ../rpmio/librpmio.so.10.2.0: undefined reference to `tan'   
/usr/bin/ld: ../rpmio/librpmio.so.10.2.0: undefined reference to `cos'   
/usr/bin/ld: ../rpmio/librpmio.so.10.2.0: undefined reference to `log'   
/usr/bin/ld: ../rpmio/librpmio.so.10.2.0: undefined reference to `pow'   
/usr/bin/ld: ../rpmio/librpmio.so.10.2.0: undefined reference to `log10' 
/usr/bin/ld: ../rpmio/librpmio.so.10.2.0: undefined reference to `sqrt' 
```

My building environment:
```
PRETTY_NAME="Ubuntu 24.04 LTS"
NAME="Ubuntu"
VERSION_ID="24.04"
VERSION="24.04 LTS (Noble Numbat)"
VERSION_CODENAME=noble
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=noble
LOGO=ubuntu-logo
```

My installations:
```
apt install pkg-config
apt install libbz2-dev
apt install libpopt-dev
apt install libreadline-dev
apt install libzstd-dev
apt install libelf-dev
apt install libdw-dev
apt install libarchive-dev
apt install libmagic-dev
apt install libsqlite3-dev
apt install python3-dev
apt install libcap-dev
apt install libaudit-dev
apt install libselinux-dev
apt install libgcrypt1-dev
apt install libdbus-1-dev
apt install podman
apt install docker.io
```

**Comment by iseki0** [comment]
Before I post here, I tried to add `target_link_libraries(elfdeps PRIVATE m)` and `target_link_options(elfdeps PRIVATE "LINKER:--copy-dt-needed-entries")`. But it doesn't works.

**Comment by pmatilai** [comment, marked as answer]
> /usr/bin/ld: ../rpmio/librpmio.so.10.2.0: undefined reference to `sinh' 

It's librpmio that's missing linkage to libm, elfdeps is just the messenger. Now, librpmio itself doesn't need libm for anything, it's Lua that does, and you seem to be statically linking it:
> -DLUA_LIBRARIES=/root/lua-5.4.7/src/liblua.a

We don't generally endorse or support static linkage because it's nothing but a headache. As you're witnessing here. It might work if you add -libm to librpmio linked libraries but it's equally possible there are further issues.

That /root/... path also suggests you're building as root which you absolutely should not be doing, for any software.

**Reply by iseki0** [reply]
> We don't generally endorse or support static linkage because it's nothing but a headache.

But I read the docs of lua.org, there's no guide about compile lua to a dynamic library.

**Reply by pmatilai** [reply]
All the distros build it as a dynamic library. It should also be available on Ubuntu, so you don't need to roll your own.
